### PR TITLE
Madara fixes

### PR DESCRIPTION
**Pull request by PythonCoderAS**
Small fixes I made on the Madara base class, fixing bugs.

### CODE REVIEW

**Comment by PythonCoderAS** [comment]
This change was made because some sources don't have enough manga to signal that pagination is finished, so Madara will request page 2 and so on, which are completely empty. If page 2 is empty, that for some reason signals to go onto page 3, and so on. This checks if the response is empty, and returns if it is.

https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/Madara.ts#L175-L177

**Comment by PythonCoderAS** [comment]
https://github.com/Paperback-iOS/extensions-madara/pull/29/files#diff-d2e73e5111a79b7f0a4eaad66f583ba22c5c9ff08e836ce1d275f3ec2e6ede9eR15-R22

**Comment by PythonCoderAS** [comment]
Some sources, especially those that are made for 1-3 manga, don't have a method of finding the manga ID. In these cases, I override the Madara class's `getNumericID` function. These changes checks if the shortlink method of manga ID exists, and if it doesn't, a call to the source's `getNumerID` function is made. This required changing the method to async since the `getNumericID` function is async as well.
https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/MadaraParser.ts#L15-L22

**Comment by PythonCoderAS** [comment]
Some sources uses the `row` class to denote chapters rather than `wp-manga-chapter` class. I also moved the selector to not duplicate calls.
https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/MadaraParser.ts#L70

**Comment by PythonCoderAS** [comment]
The variable `should = chai.should()` is unused, so I used find and replace to delete it.

**Comment by GameFuzzy** [comment]
> Some sources uses the `row` class to denote chapters rather than `wp-manga-chapter` class. I also moved the selector to not duplicate calls.
> 
> https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/MadaraParser.ts#L70



> Some sources uses the `row` class to denote chapters rather than `wp-manga-chapter` class. I also moved the selector to not duplicate calls.
> 
> https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/MadaraParser.ts#L70

Abstract this out

**Comment by GameFuzzy** [comment]
> The variable `should = chai.should()` is unused, so I used find and replace to delete it.

This is just unnecessary imo.

**Comment by GameFuzzy** [comment]
> This change was made because some sources don't have enough manga to signal that pagination is finished, so Madara will request page 2 and so on, which are completely empty. If page 2 is empty, that for some reason signals to go onto page 3, and so on. This checks if the response is empty, and returns if it is.
> 
> https://github.com/Paperback-iOS/extensions-madara/blob/498002294d3c04f04a4b34c366d4e0beee9c58e1/src/Madara.ts#L175-L177

Should be looked over by @nar1n

**Comment by PythonCoderAS** [comment]
> > The variable `should = chai.should()` is unused, so I used find and replace to delete it.
> 
> This is just unnecessary imo.

It is but I don't see why there should be another PR to get rid of them. Also my IDE complains about "unused variables" when I leave them there.

**Comment by GameFuzzy** [comment]
> > > The variable `should = chai.should()` is unused, so I used find and replace to delete it.
> > 
> > 
> > This is just unnecessary imo.
> 
> It is but I don't see why there should be another PR to get rid of them. Also my IDE complains about "unused variables" when I leave them there.

It’s like that in every repo. Any test file should be able to just be copied and used for another source with few modifications. If someone wants to use `should` they shouldn’t need to import it.

**Comment by PythonCoderAS** [comment]
All tests use `except` rather than should. Using should requires a bunch of the testing code to change regardless, since it's all written in terms of expect.

**Comment by GameFuzzy** [comment]
> All tests use `except` rather than should. Using should requires a bunch of the testing code to change regardless, since it's all written in terms of expect.

Either way, we should keep the test files the same as they are everywhere else imo.

File: src/MadaraParser.ts [hunk 1]
```diff
@@ -60,16 +67,16 @@ export class Parser {
 
     parseChapterList($: CheerioSelector, mangaId: string, source: any): Chapter[] {
         let chapters: Chapter[] = []
-
+        let selector = $('li.'+source.chapterRowClass);
```
Comment (GameFuzzy) [inline src/MadaraParser.ts:70]: Why are you doing this? If you're gonna give source control over the selector then give the source full control over the selector.

**Review by GameFuzzy — CHANGES_REQUESTED** [review]

File: src/MadaraParser.ts (hunk 1 above)
Comment (PythonCoderAS) [reply src/MadaraParser.ts:70]: I see.

File: src/MadaraParser.ts [hunk 2]
```diff
@@ -8,11 +8,18 @@ import {
     Tag,
     TagSection
 } from "paperback-extensions-common";
+import {Madara} from "./Madara";
 
 export class Parser {
 
-    parseMangaDetails($: CheerioSelector, mangaId: string): Manga {
-        let numericId = $("script#wp-manga-js-extra").get()[0].children[0].data.match('"manga_id":"(\\d+)"')[1]
+    async parseMangaDetails($: CheerioSelector, mangaId: string, source: Madara): Promise<Manga> {
```
Comment (ConradWeiser) [inline src/MadaraParser.ts:15]: Could you please comment your changes a little more extensively in the typescript files? Specifically for this method.
Parsing is very hard to read if you're a newcomer. When there's niche things like this being added in, it should have a detailed explanation of the preconditions where this logic would apply. 

This appears to be something very few sources use. As such, people trying to understand how to write a module for the grand majority of sources should have some kind of documentation informing them that this logic may not apply for them.

File: src/Madara.ts [hunk 3]
```diff
@@ -43,6 +43,11 @@ export abstract class Madara extends Source {
      */
     hasAdvancedSearchPage: boolean = false
 
+    /**
+     * Different Madara sources use a different selector to denote chapter rows.
```
Comment (ConradWeiser) [inline src/Madara.ts:47]: Elaborate a little more on this. What is a chapter row? What is the container `div` or such, which this is found in?
If I knew nothing about madara, this comment should help guide me to determine what this selector should be used for the source that I was implementing.

**Review by ConradWeiser — CHANGES_REQUESTED** [review]
I don't have too much of an issue with these changes. Sure, it's an adaptation towards a niche subset of Madara sources, but I don't see evidence that it'll break current Madara sources.

Provided that the documentation is detailed out a little more, I'd be comfortable doing a rigor test in-app, and approving things if it works out.

**Review by ConradWeiser — APPROVED** [review]
Very good documentation! Thanks!

**Review by GameFuzzy — APPROVED** [review]
yup, looks good

**Pull request merged**